model/riscv_sys_control.sail

```diff
@@ -408,8 +482,187 @@ function handle_exception(e: ExceptionType) -> unit = {
   set_next_pc(exception_handler(cur_privilege, CTL_TRAP(t), PC))
 }
 
-function handle_interrupt(i : InterruptType, del_priv : Privilege) -> unit =
+function handle_interrupt(i : InterruptType, del_priv : Privilege) -> unit = {
+  IEtrigger_match(cur_privilege, num_of_interruptType(i), false);
   set_next_pc(trap_handler(del_priv, true, interruptType_to_bits(i), PC, None(), None()))
+}
+
+function find_first_zero(v: xlenbits) -> {'n, (1 <= 'n < xlen). int('n)} = {
+  foreach (i from 1 to (xlen - 1)) {
+    if v[i .. i] == 0b0 then return i;
+  };
+  return 1;
+}
+
+/* Match Control Type 6 Trigger Match */
+function instrDataMatch(cur_priv : Privilege, addr : xlenbits, data : xlenbits, match_size : matchSize, match_type : MatchType) -> bool = {
+  var raise_exception : bool = false;
+  let half_xlen = sizeof(half_xlen);
+  /* Traverse through all triggers */
+  foreach (i from 0 to (sizeof(N_TRIGGERS) - 1)) {
+    let trigger_type : bits(4) = tdata1_x.tdata1[i][(xlen - 1) .. (xlen - 4)];
+    var data_matched : bool = false;
+    var size_matched : bool = false;
+    /* Match Control Type 6 Trigger found */
+    if (trigger_type == triggerType_to_bits(TRIG_MCONTROL6)) then {
+      var MCtrl6 : MatchControlType6  = Mk_MatchControlType6(tdata1_x.tdata1[i][26 .. 0]);
+      let matchOption : bits(4) = MCtrl6[match_opt];
+      /* Decide content to match address/data based on select bit */
+      let tdata2_content : xlenbits = match (MCtrl6[select]) {
+        0b0 => { size_matched = true; addr },
+        0b1 => {
+          size_matched = match matchSize_of_bits(MCtrl6[size]) {
+            ANY => true,
+            _   => (match_size == matchSize_of_bits(MCtrl6[size])),
+          };
+          data
+        }
+      };
+      /* Data matched according to the select match options */
+      data_matched = match (matchOpt_of_bits(matchOption)) {
+        MOPT_EQUAL => { (tdata2_x.tdata2[i] == tdata2_content) },
+        MOPT_NAPOT => {
+          let idx : {'n, (1 <= 'n < xlen). int('n)} = find_first_zero(tdata2_x.tdata2[i]);
+          tdata2_x.tdata2[i][(xlen - 1) .. idx] == tdata2_content[(xlen - 1) .. idx]
```

Comment:
Shouldn't this be (idx+1) in both places? You don't want to include that 0 in the comparison. Per the spec, the top M bits are compared, and "M is XLEN-1 minus the index of the least-significant bit containing 0 in tdata2".

Assuming ranges are inclusive, here you are comparing (xlen-1)-idx+1 bits.

Reply:
Fixed